usr/src/lib/libzfs/common/libzfs_sendrecv.c

```diff
@@ -609,11 +621,23 @@ send_iterate_snap(zfs_handle_t *zhp, void *arg)
 {
 	send_data_t *sd = arg;
 	uint64_t guid = zhp->zfs_dmustats.dds_guid;
+	uint64_t txg = zhp->zfs_dmustats.dds_creation_txg;
 	char *snapname;
 	nvlist_t *nv;
 
 	snapname = strrchr(zhp->zfs_name, '@')+1;
 
+	if (sd->parent_tosnap_txg != 0 && txg > sd->parent_tosnap_txg) {
```

Comment:
Sure. Please let me explain the motivation of this change, using an example:

Before replication, during the traversal of tree of datasets, new algorithm can skip (without errors) nested datasets under two conditions:

Nested dataset **A/B** does not contain a snapshot **tosnap** **AND** dataset **A/B** was created later than the snapshot **tosnap** of the parent dataset **A**.

But as before, new algorithm returns an error in the case where the nested dataset **A/B** does not contain a snapshot **tosnap** **AND** dataset **A/B** was created earlier than the snapshot **tosnap** for its parent dataset **A**.

When I started testing this change, I realized that the replication stream header includes a lot of redundant information - details of the snapshots that were created after the snapshot **tosnap**. Therefore, I decided to use the approach described above while traversing snapshots. This change is based on a comparison of TXG’s, reduces overhead when filling the header for the replication stream. This is especially noticeable on systems containing a large number of snapshots.

**Example**:

dataset A
```
     snap1   snap2   snap3   snap4   snap5   snap6   snap7   snap8
_______|_______|_______|_______|_______|_______|_______|_______|___________ >
       x1      x2      x3      x4      x5      x6      x7      x8             txg

```

dataset A/B
```
         snap1   snap2   snap3   snap4   snap5   snap6   snap7   snap8
___________|_______|_______|_______|_______|_______|_______|_______|_______ >
           y1      y2      y3      y4      y5      y6      y7      y8         txg

```

initial replication: `# zfs send -R A@snap1`:
```
send_iterate_fs() run for A and A/B:
	for A:
		parent_fromsnap_txg == 0
		parent_tosnap_txg   == x1
		so we can skip all snapshots with create_txg > x1

	for A/B:
		parent_fromsnap_txg == 0
		parent_tosnap_txg   == y1
		so we can skip all snapshots with create_txg > y1

send_iterate_snap() run for every snapshot:
	for A: add snap1 into send stream header and skip snapshots snap2-snap8
	(because its txg’s > parent_tosnap_txg)

	for A/B: add snap1 into send stream header and skip snapshots snap2-snap8
	(because its txg’s > parent_tosnap_txg)

```
This means that we do not include information about snapshots snap2-snap8 into the replication stream header, in case of replication snapshot snap1

**Simple test case** with additional debug code:
```
--- a/usr/src/lib/libzfs/common/libzfs_sendrecv.c
+++ b/usr/src/lib/libzfs/common/libzfs_sendrecv.c
@@ -1624,6 +1624,7 @@ zfs_send(zfs_handle_t *zhp, const char *fromsnap, const char *tosnap,
                        VERIFY(0 == nvlist_add_nvlist(hdrnv, "fss", fss));
                        err = nvlist_pack(hdrnv, &packbuf, &buflen,
                            NV_ENCODE_XDR, 0);
+                       (void) fprintf(stderr, "nvlist_pack: %u bytes\n", buflen);
                        if (debugnvp)
                                *debugnvp = hdrnv;
                        else
```

**Steps**:
```
# for i in {1..10}; do for j in {1..10}; do zfs create -p tank/src/ds$i/nested$j; done; done
# for i in {1..1000}; do zfs snapshot -r tank/src@snap$i; done
```

**Old code result**:
```
# ptime zfs send -R tank/src@snap1 | zfs recv tank/dst
nvlist_pack: 8911612 bytes

real     4:06.012471011
user       27.129523402
sys      2:44.492517170
```

**New code result**:
```
# ptime zfs send -R tank/src@snap1 | zfs recv tank/dst
nvlist_pack: 40492 bytes

real     2:26.751300870
user        9.152301372
sys      1:28.655915062
```

Thank you!